### PR TITLE
Interop: local devnet

### LINKED ISSUE #11306: Interop: local devnet
As part of devnet 1, we will start with a local docker-compose, to run the full stack.

Unlike op-e2e, this is not as tightly coupled to the current deployment flow, and unblocks runing the interop stack.

For the local devnet we have several options:
1. Extend the current monorepo devnet
2. Reuse some of the current monorepo devnet, but provide a separate entrypoint
3. Copy-paste some of the monorepo devnet
4. Fork of the devnet as a separate repository that deploys the stack

Currently I believe 2 is the best approach, as this can run in CI, while on the same `develop` branch.

Something similar was implemented here: https://github.com/ethereum-optimism/optimism/pull/10777
However, this targeted an older version of the devnet (before L1 beacon node) and deploys duplicate Superchain contracts to the L1, and misconfigures the addresses of the L2s and their L1 deployments.

We can fork this work, modify the deployment, and then integrate the op-supervisor to unblock interop testing.

The deploy-code should be modified to reuse this for a live long-running devnet.

The primary blocker on getting the devnet running is the genesis-setup. See https://github.com/ethereum-optimism/optimism/issues/11493

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

Local docker-compose to run an interop devnet.

Builds on top of https://github.com/ethereum-optimism/optimism/pull/11589

TODO:
- [x] implement CLI around `interopgen` package to build the genesis / config data needed
- [x] ~adapt docker compose or entry-points, to configure sequencer/proposer/batcher/etc. keys using interop devkeys package.~ generate env files with keys retrieved from devkey CLI.

**Tests**

Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.

**Metadata**

Fix https://github.com/ethereum-optimism/optimism/issues/11306

### CODE REVIEW

**Comment by protolambda** [comment]
Ready for review, but do not merge yet, as this is stacked on top of the interop-gen PR, which is not in a state where it can be merged into develop yet (due to pre-OPSM deploy workarounds).

File: interop-devnet/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,343 @@
+# This Compose file is expected to be used with the devnet-up.sh script.
+# The volumes below mount the configs generated by the script into each
+# service.
```
Comment (axelKingsley) [inline interop-devnet/docker-compose.yml:3]: is it the case it's expected to be used with `devnet-up`? I don't see any modifications to that script so I wanted to make sure this wasn't a copy-paste error.

**Review by axelKingsley — DISMISSED** [review]
Generally looks good. Left optimistic Approval based on - did you run this locally? I assume so, and I assume it created the appropriate entities.

**Comment by protolambda** [comment]
Sorry for the notification noise y'all. Big rebase, change of base branch, ended up with `develop` commits temporarily in this PR diff.

**Comment by protolambda** [comment]
Rebase complete, but need to test run it with latest changes. Will work on that after sorting our some remaining work on the interop deposits side.

**Comment by protolambda** [comment]
Rebased on latest `interopgen2`. Will try and testrun this tomorrow, but think the op-e2e tests are more productive first, since there are some quirks with safety progression of op-node with empty op-supervisor DB, as found by @axelKingsley

File: .gitignore [hunk 2]
```diff
@@ -36,6 +36,8 @@ packages/contracts-bedrock/deployments/anvil
 
 .devnet*
 
+.devnet-interop
+
```
Comment (ghost) [inline .gitignore]: Is this not covered by the above `.devnet*` entry?

File: interop-devnet/create-chains.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+chain_env=".devnet-interop/env/l2/900201"
+mkdir -p "$chain_env"
+key_cmd="go run ./op-node/cmd interop devkey secret --domain=chain-operator --chainid=900201"
+# op-node
+echo "OP_NODE_P2P_SEQUENCER_KEY=$($key_cmd --name=sequencer-p2p)" >> "$chain_env/op-node.env"
+# proposer
+echo "OP_PROPOSER_PRIVATE_KEY=$($key_cmd --name=proposer)" >> "$chain_env/op-proposer.env"
+echo "OP_PROPOSER_GAME_FACTORY_ADDRESS=$(jq -r .DisputeGameFactoryProxy .devnet-interop/deployments/l2/900201/addresses.json)" >> "$chain_env/op-proposer.env"
+# batcher
+echo "OP_BATCHER_PRIVATE_KEY=$($key_cmd --name=batcher)" >> "$chain_env/op-batcher.env"
```
Comment (axelKingsley) [inline interop-devnet/create-chains.sh:75]: An opportunity to DRY if we made this a function. But for N=2 this is not a blocker

File: interop-devnet/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    ports:
+      - "8545:8545"
+      - "8546:8546"
+      - "7060:6060"
+    volumes:
+      - "l1_data:/db"
+      - "${PWD}/../.devnet-interop/genesis/l1/genesis.json:/genesis.json"
+      - "${PWD}/../ops-bedrock/test-jwt-secret.txt:/config/jwt-secret.txt"
+    environment:
+      GETH_MINER_RECOMMIT: 100ms
```
Comment (axelKingsley) [inline interop-devnet/docker-compose.yml:32]: Do we really need a different recommit value than the default?

File: interop-devnet/justfile [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  # Note: `justfile` interprets the curly brackets. So we escape them, by wrapping it with more, as a string, like Jinja2.
+  docker image ls 'interop-devnet*' --format='{{ '{{.Repository}}' }}' | xargs -r docker rmi
+  docker volume ls --filter name=interop-devnet --format='{{ '{{.Name}}' }}' | xargs -r docker volume rm
+
+devnet-logs:
+  docker compose logs -f
+
+supervisor-workaround:
+  cast rpc --rpc-url="http://localhost:9045" admin_addL2RPC "ws://l2-a:8546"
+  cast rpc --rpc-url="http://localhost:9045" admin_addL2RPC "ws://l2-b:8546"
```
Comment (axelKingsley) [inline interop-devnet/justfile]: We don't need this workaround any more. Now that the Supervisor is accepting of not having an L2 connection for chains, the supervisor can be stood up, the RPC be determined, and then the URL be used in the OP Node start options.

File: op-node/cmd/interop/interop.go [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+				if err := writeJson(filepath.Join(l2Dir, "rollup.json"), dep.RollupCfg); err != nil {
+					return fmt.Errorf("failed to write L2 %s rollup config: %w", id, err)
+				}
+			}
+		}
+		return nil
+	},
+}
+
+func writeJson(path string, content any) error {
```
Comment (axelKingsley) [inline op-node/cmd/interop/interop.go:174]: I'm surprised we don't have some utility for this already in the monorepo? Could we chuck this into a utility package?

**Review by axelKingsley — APPROVED** [review]
Changes all make sense! I left a couple optional comments.

File: op-node/cmd/interop/interop.go (hunk 6 above)
Comment (protolambda) [reply op-node/cmd/interop/interop.go:174]: there is, but it was incomplete, it doesn't allow you to just write a basic file. added some missing utils, and now sharing the same op-service json-writing code.

File: interop-devnet/justfile (hunk 5 above)
Comment (protolambda) [reply interop-devnet/justfile]: removed the workaround

File: interop-devnet/docker-compose.yml (hunk 4 above)
Comment (protolambda) [reply interop-devnet/docker-compose.yml:32]: yes, with fast block times you need this for txs to be included when submitted mid-way through the block time window.

File: .gitignore (hunk 2 above)
Comment (protolambda) [reply .gitignore]: good catch

File: interop-devnet/create-chains.sh [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  --artifacts-dir=packages/contracts-bedrock/forge-artifacts \
+  --foundry-dir=packages/contracts-bedrock \
+  --l1.chainid=900100 \
+  --l2.chainids=900200,900201 \
+  --out-dir=".devnet-interop" \
+  --log.format=logfmt \
+  --log.level=info
+
+# create L1 CL genesis
+eth2-testnet-genesis deneb \
```
Comment (axelKingsley) [inline interop-devnet/create-chains.sh:37]: The root level Makefile has handling to install this tool if it's not found, as it was not in my case. We could consider doing the same.

**Review by axelKingsley — COMMENTED** [review]
(moving my in-progress assessment to the next comment)

**Comment by axelKingsley** [comment]
On the latest runs, I see Unsafe, Safe and Finalized update when and as needed. Recent clean runs have connected without issue and make the progress expected from the network. On occasion I've seen issues where the Finalized head can't be fetched from the Supervisor, but I believe the latest fixes have eliminated this.

**Pull request merged**